### ISSUE TITLE
feat: implement ASR

### ISSUE THREAD

**Issue by Randomidous** [snesmaeili/PyZapline_plus#20]
**Describe the solution you'd like**
Pick up implementation of ASR and continuous-DSS for inclusion in the package.

_Original work:_
- [meegkit's ASR](https://github.com/nbara/python-meegkit/blob/master/meegkit/asr.py)
- [seminal ASR paper](https://sccn.ucsd.edu/githubwiki/files/asr-final-export.pdf)
- [presentation on ASR](https://sccn.ucsd.edu/githubwiki/files/asr-final-export.pdf)
- [matlab implementation](https://eeglab.org/plugins/clean_rawdata/)

_Adaptations:_
- [Adaptive ASR (deprecated)](https://github.com/t5i0m7/AASR) 
- [Riemannian ASR (deprecated)](https://github.com/s4rify/rASRMatlab)
- [PyASR (closed)](https://github.com/mne-tools/mne-python/pull/9302)
- [Juggler's ASR](https://pubmed.ncbi.nlm.nih.gov/40324599/)

**Describe the solution you'd like**
A robust, quick, and tested implementation of the Artifact Subspace Reconstruction method

**Comment by payamsash** [comment]
Hi all, jumping in here from [MNE-RT](https://github.com/payamsash/mne-rt)
I ran into multiple issues with the implementations and ended up writing new one. The design difference is that mine is a calibrate-once / apply-per-chunk streaming implementation rather than a batch processor:

```
asr = ASRDenoiser(cutoff=5, win_len=0.5)
asr.fit(baseline_raw)           # runs once
clean_chunk = asr.transform(chunk)   # called every window on data shape (n_channels, n_samples)
```
 
`fit()` computes the calibration covariance and per-component RMS thresholds from a baseline `Raw`. `transform()` applies the projector to one incoming chunk with no memory of past or future, which is fast enough for real-time use and avoids holding the full recording in memory.

Relevance to this issue as suggested by @larsoner, if mne-denoise adopted a fit(raw_or_ndarray) / transform(ndarray) interface alongside the existing batch API, the same implementation could be used both offline and online without duplication.

**Comment by larsoner** [comment]
+1 for a sklearn-like framework under the hood. It should be easy(ish) in principle to make this work for raw/epochs/evoked just like we do for ICA in MNE-Python.

@payamsash would you be up for moving your code over here and implementing this if @BabaSanfour and others are on board?

**Comment by BabaSanfour** [comment]
Hi @payamsash (and @larsoner ).. could you specify which implementations you checked for ASR? we ran into issues with the existing ones too that is why we have been working on a native implementation for ASR (#36 #40 #41 ) it mainly led by sina and tested / reviewed by me + some lab members for more alpha testing.. all our denoisers are sklearn like with fit and transform .. we aim to ship this by end / beginning of next week. What we can do is we finish the work.. you try to test it with your codebase and then add the features / additions you already implemented. how does this sound ? .. if you have time you can even look at the current state of the implementation and suggest / add things.

**Comment by BabaSanfour** [comment]
Hi again @payamsash and @larsoner,

We’ve made good progress on the ASR implementation and are now at a point where we’re quite happy with it for a first release.

At the moment, we support four ways of using ASR, including an `AdaptiveASR` class that is fairly close to what you described. Its `.transform()` method processes data chunk-by-chunk without loading the entire recording into memory, so it addresses the memory issue you mentioned. Internally, it keeps only a very small state containing the edges of the previous chunk. This allows proper overlap-add blending across chunk boundaries, keeping the reconstructed signal smooth and continuous. One important difference, though, is that `AdaptiveASR` continuously updates its calibration model over time, whereas your proposal follows a static "calibrate once, then apply" strategy.

The approach you suggested would effectively be a stateless "calibrate-once / apply-per-chunk" mode, which avoids the overlap-add step altogether. From a computational standpoint, this is extremely efficient and very attractive for low-latency applications such as BCIs. The main tradeoff is on the signal-processing side: applying different reconstruction matrices independently to successive chunks can introduce discontinuities at chunk boundaries. Those discontinuities create artificial high-frequency components (spectral leakage), which may affect downstream frequency-domain analyses.

That said, I completely agree that for many real-time applications, users may be willing to accept this tradeoff in exchange for maximum speed and minimal latency. So I think there is definitely value in supporting such a mode, as long as the limitations are clearly documented. @larsoner, does that sound reasonable to you?

As a next step, I think we should first benchmark both approaches: the static approach you described versus the adaptive implementation we currently have. If the runtime difference turns out to be small, I would lean toward recommending the adaptive approach directly. If the speedup is substantial and depending on @larsoner answer, then it could make sense to expose a dedicated `.transform_chunk(chunk: np.ndarray)` method on either `ASR` or `AdaptiveASR` (name and location to be determined). If that's something you'd be interested in contributing via a PR to #36 , we'd be happy to discuss it further. Let us know what you think.

**Comment by larsoner** [comment]
That sounds reasonable to me. It probably makes sense to have some way to have adaptive mode on and off. In cases of data dropouts and/or when you know data won't be continuous (only sampled for some windows of time) you'd want to disable the adaptive (at least overlap-add processing) bits, and in cases where non-adaptive is "good enough" the speedup might be worth it, too.

I took a quick look at #36 but the PR adds 23k lines (😱) so ... tough to review quickly how the code is used. But the top-comment classes splitting `ASR` (with different methods) plus `AdaptiveASR` looks reasonable to me. My 2c / drive-by comment there would be that I'm assuming the lines were generated by AI/LLM, I would ask it to try to make things much more compact (like by a factor of 10) ... I'd expect a PR like that to be closer to 2k lines which would be much more maintainable and readable than 23k lines.

**Comment by BabaSanfour** [comment]
The adaptive mode has to be called specifically, otherwise we have the normal (non-adaptive) ASR. The question is about for adaptive do we have stateless "calibrate-once / apply-per-chunk" mode that @payamsash suggested or what we have now for adaptive: keeps only a very small state containing the edges of the previous chunk which helps continuously updates its calibration model over time. But I think ideally we would benchmark both and how much compute they require. 

as for the PR, yes 23k is a lot but its purely scripts that we used to benchmark our methods compared to other ASR implementations/matlab + documentation; most of these are generated by LLMs and hosted in scripts not in the main mne-denoise. I can move them to another branch.. but the idea was to remove them once we finish all the modules we are targeting and maybe start working on a paper with release 0.0.2. 

So the actual code (mne_denoise.asr is only 4k line of code) and this is because we worked on ASR for a while in many locations and then merged everything to this branch; and it also has 4-5 methods not just one, each requiring its own helpers. The core code was with the help of LLMs also (for a first phase of design )but I read every line of code , did a lot of modifications and the final structure; codes are ours not just LLM.

**Comment by BabaSanfour** [comment]
@payamsash just pinging you if you forgot to respond to this. we want to move on and merge this PR. If you don't have time to review / make the contribution, we understand :D you are always welcome to open a new issue in the future.

**Comment by payamsash** [comment]
Hi @BabaSanfour, sorry for my late reply. I had a look at your implementation and always having real-time scenarios in my mind, I have couple of suggestions: 
1. in rt when we have chunk of data the format is numpy array (at least in `mne-rt`), so having sth like `transform_chunk(chunk: np.ndarray) -> np.ndarray` which skips MNE round-trip would be nice.
2. The `AdaptiveASR` class is continuous over chunks, but in `ASR` class we only have the option of chunk-to-chunk. We only have one calibration in `ASR`, but I think it's ok to have continues chunks here as well. no?

**Comment by BabaSanfour** [comment]
Hi @payamsash

1. actually, our current `transform` method actually already accepts `np.ndarray` and bypasses the MNE round-trip overhead internally (it just does a quick `isinstance` check)! 

Here is the signature in `mne_denoise/asr/core.py`:
```python
    def transform(
        self,
        X: BaseRaw | BaseEpochs | Evoked | np.ndarray,
        y=None,
        copy: bool | None = None,
        return_diagnostics: bool = False,
    ) -> Any:
```
2. for continuous chunks for the standard `ASR` class: as mentioned before you could benchmark the compute overhead of `AdaptiveASR`. If the overhead is negligible, `AdaptiveASR` directly for continuous chunking would be better. For use cases, over long real-time sessions, an adaptive baseline is actually safer than a static one because it tracks slow environmental changes (like gel drying or cap shifts) that would otherwise invalidate a static calibration. By keeping a slow learning rate, `AdaptiveASR` already gives you smooth chunk boundaries while safely handling these physiological drifts.

**Comment by BabaSanfour** [comment]
Closing this issue as we merged ASR; thanks everyone for contributing.. for new discussions on ASR we can add a new issue.
